File: environment-spring-boot/casadocodigo/src/main/java/org/ecorp/casadocodigo/services/CategoriaService.java
package org.ecorp.casadocodigo.services;

import java.util.Optional;
import javax.validation.Valid;
import org.ecorp.casadocodigo.dtos.CategoriaDTO;
import org.ecorp.casadocodigo.exceptions.ResourceNotFoundException;
import org.ecorp.casadocodigo.forms.CategoriaFormRequest;
import org.ecorp.casadocodigo.model.Categoria;
import org.ecorp.casadocodigo.repositories.CategoriaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class CategoriaService {


  @Autowired
  private CategoriaRepository repository;

  public CategoriaDTO create(@Valid CategoriaFormRequest categoriaForm) {

    Categoria saved = repository.save(categoriaForm.map());
    return new CategoriaDTO(saved);

  }

  public CategoriaDTO buscaCategoriaById(Long id) {

    Optional<Categoria> found = repository.findById(id);
    if (found.isEmpty()) {
      throw new ResourceNotFoundException();
    }
    return new CategoriaDTO(found.get());
  }

}

File: src/main/kotlin/com/tiozao/cdd/loja/domain/service/CategoriaService.kt
package com.tiozao.cdd.loja.domain.service

import com.tiozao.cdd.loja.domain.model.CategoriaModel
import com.tiozao.cdd.loja.repository.entity.CategoriaEntity
import com.tiozao.cdd.loja.repository.CategoriaRepository
import com.tiozao.cdd.loja.repository.extensions.toEntity
import com.tiozao.cdd.loja.repository.extensions.toModel
import javax.validation.Valid
import org.springframework.stereotype.Service
import org.springframework.validation.annotation.Validated

@Service
@Validated
class CategoriaService(private var repository: CategoriaRepository) {

    fun createCategoria(@Valid categoria: CategoriaModel): CategoriaModel {
        return repository.save(
            categoria.toEntity())
            .toModel()
    }

    fun findCategoria(categoriaId: Int) =
         repository.findById(categoriaId).orElse( null)

}